hystrix-contrib/hystrix-javanica/hystrix-javanica-core/src/main/java/com/netflix/hystrix/contrib/javanica/command/MetaHolder.java

```diff
@@ -15,96 +15,94 @@
-@Immutable
-public final class MetaHolder {
-
-    private final HystrixCollapser hystrixCollapser;
-    private final HystrixCommand hystrixCommand;
-    private final DefaultProperties defaultProperties;
-
-    private final Method method;
-    private final Method cacheKeyMethod;
-    private final Method ajcMethod;
-    private final Method fallbackMethod;
-    private final Object obj;
-    private final Object proxyObj;
-    private final Object[] args;
-    private final Closure closure;
-    private final String defaultGroupKey;
-    private final String defaultCommandKey;
-    private final String defaultCollapserKey;
-    private final String defaultThreadPoolKey;
-    private final ExecutionType executionType;
-    private final boolean extendedFallback;
-    private final ExecutionType collapserExecutionType;
-    private final ExecutionType fallbackExecutionType;
-    private final boolean fallback;
-    private boolean extendedParentFallback;
-    private final boolean defaultFallback;
-    private final JoinPoint joinPoint;
-    private final boolean observable;
-    private final ObservableExecutionMode observableExecutionMode;
-
-    private static final Function identityFun = new Function<Object, Object>() {
-        @Nullable
-        @Override
-        public Object apply(@Nullable Object input) {
-            return input;
-        }
-    };
-
-    private MetaHolder(Builder builder) {
-        this.hystrixCommand = builder.hystrixCommand;
-        this.method = builder.method;
-        this.cacheKeyMethod = builder.cacheKeyMethod;
-        this.fallbackMethod = builder.fallbackMethod;
-        this.ajcMethod = builder.ajcMethod;
-        this.obj = builder.obj;
-        this.proxyObj = builder.proxyObj;
-        this.args = builder.args;
-        this.closure = builder.closure;
-        this.defaultGroupKey = builder.defaultGroupKey;
-        this.defaultCommandKey = builder.defaultCommandKey;
-        this.defaultThreadPoolKey = builder.defaultThreadPoolKey;
-        this.defaultCollapserKey = builder.defaultCollapserKey;
-        this.defaultProperties = builder.defaultProperties;
-        this.hystrixCollapser = builder.hystrixCollapser;
-        this.executionType = builder.executionType;
-        this.collapserExecutionType = builder.collapserExecutionType;
-        this.fallbackExecutionType = builder.fallbackExecutionType;
-        this.joinPoint = builder.joinPoint;
-        this.extendedFallback = builder.extendedFallback;
-        this.defaultFallback = builder.defaultFallback;
-        this.fallback = builder.fallback;
-        this.extendedParentFallback = builder.extendedParentFallback;
-        this.observable = builder.observable;
-        this.observableExecutionMode = builder.observableExecutionMode;
-    }
-
-    public static Builder builder() {
-        return new Builder();
-    }
+public abstract class MetaHolder<M extends MetaHolder<M,B>,B extends MetaHolder.Builder<M, B>> {
```

Comment:
Agree. So what do you suggest in the example scenario I have given below? **Please note that `AbstractCommandMetaHolderBuilder` is not a subclass of MetaHolder.Builder**. It is another class which accepts `MetaHolder.Builder` as a constructor argument, set default properties and then create a MetaHolder and the one which you suggested to rename to `AbstractMetaHolderFactory`.

```java
public abstract class AbstractCommandMetaHolderBuilder<T extends MetaHolder, V extends MetaHolder.Builder<T,V>> {
    protected final V builder;  // MetaHolder.Builder
    protected AbstractCommandMetaHolderBuilder(V builder){
        this.builder = builder;
    }
    public T build() {
        builder.args(args);
	builder.method(method);
	builder.obj(obj);
        // set other properties here

        // Should I cast to T like 'return (T)builder.build()'
         return builder.build();
    }
}
```